app/models/king.rb

```diff
@@ -1,7 +1,68 @@
+# rubocop:disable Metrics/AbcSize
 class King < Piece
   def move_legal?(to_row, to_col)
     row_diff = (row - to_row).abs
     col_diff = (col - to_col).abs
     row_diff <= 1 && col_diff <= 1
+    # add castling moves to the 'whitelist', then verify the validity of the castle
+    #   further down in the logic
+  end
+
+  # override to enable castling
+  def finalize_move!(to_row, to_col)
+    return false if get_attackers(to_row, to_col).present?
+    move_to!(to_row, to_col)
+    # for castling something like: move_the_rook if castling?
+  end
+
+  def in_check?
+    get_attackers(row, col).present?
+  end
+
+  # call this method on the enemy king after the end of a turn
+  def checkmate?
+    return false unless in_check? && valid_moves.empty?
+    attackers = get_attackers(row, col)
+    return true if attackers.length > 1
+    return false if attacker_is_vulnerable?(attackers.first)
+    return false if attacker_is_blockable?(attackers.first)
+    true
+  end
+
+  def stalemate?
+    !in_check? && valid_moves.empty? && allies.all? { |ally| ally.cant_move? }
+  end
+
+  def move_space
+    [[row + 1, col + 1], [row + 1, col], [row + 1, col - 1],
+     [row,     col + 1],                 [row,     col - 1],
+     [row - 1, col + 1], [row - 1, col], [row - 1, col - 1]]
+  end
+
+  def valid_moves
+    move_space.select { |move| valid_move?(move[0], move[1]) }.select do |move|
+      get_attackers(move[0], move[1]).empty?
+    end
+  end
+
+  def get_attackers(check_row, check_col)
+    game.pieces.where.not(is_black: is_black).select do |enemy|
+      enemy.valid_move?(check_row, check_col)
+    end
+  end
+
+  def attacker_is_vulnerable?(attacker)
+    allies.any? { |ally| ally.valid_move?(attacker.row, attacker.col) }
```

Comment:
**Edit:** actually I think this is what `self_check?` is checking so this may already be covered.  I'll leave the diagram below because it looks pretty.  😛 

`self_check?` may need to be integrated with `valid_move?` somehow though...

----

Actually, maybe the `valid_move?` of every piece should check to ensure that the move doesn't place your own king into check.

I believe this scenario would not currently register as checkmate:

<img width="350" alt="screenshot 2017-04-10 17 28 52" src="https://cloud.githubusercontent.com/assets/130874/24885280/36af62ca-1e13-11e7-8031-9d73ca90db59.png">

The black queen can't actually move to capture the rook because she's blocking the white bishop.  I think just verifying that any piece movement doesn't cause check should fix this.  (and it's a rule anyways, even outside of checkmate)